### DISCUSSION TITLE
Asynchronous events - Next steps?

### DISCUSSION THREAD

**Discussion by arthuro555** [4ian/GDevelop#3623] [Enhancement ideas (and their technical discussions)]
With #3535 coming close to being merged, the next steps need to be clarified. This thread is meant both as a To-Do list for everything asynchronous and to discuss obscure points of it.

To-do:

- [x] Add hot reloading of callbacks
- [ ] Add `Promise.all` equivalent
- [ ] Add `Promise.any` equivalent
- [ ] Add `Promise.race` equivalent
- [x] Allow making asynchronous actions in events functions extensions
- [ ] Convert existing extensions to asynchronous
  - [x] Network requests
  - [ ] Firebase
  - [ ] P2P - In Progress: #3778 
  - [x] Filesystem - Done: #5001
  - [x] Clipboard
- [ ] Add asynchronous expressions?
- [x] Add actions with toggelable awaiting

**Comment by 4ian** [comment]
Just to state it here:

> Toggelable asynchronicity

Agree. We'll have most of them optionnally async and some of them (like a "Wait actions") forced to be async.

**Reply by arthuro555** [reply]
How should we implement those optionally async functions, since the async implementations is often pretty different from the synchronous one? I think we should have a `setAsyncFunction` method for optional asynchrony that will add another function name for the async version. An action purely asynchronous could also just use `setAsyncFunction` without `setFunction` instead of our current `setAsync` to make the API more concise.

**Comment by arthuro555** [comment]
Progressing the goals with #5001
